### ISSUE TITLE
`iso` schemable

### ISSUE THREAD

**Issue by skeate** [jacob-alford/schemata-ts#151]
Idea is to generalize all these type-conversion additions.

`iso` would be an addition to the `Schemable` interface that takes an isomorphism `A<->B` where `A` is a narrowed (`refine`d) version of the general type and `B` is the target type. For example:

```ts
const DateFromInt = pipe(
  Int(),
  Refine(isValidDateInt),
  Iso(dateIntIso),
)

const isValidDateInt = (n: Int): n is DateInt =>
  !Number.isNaN((new Date(Number.MAX_SAFE_INTEGER)).getTime())
declare const dateIntIso: Iso<DateInt, Date>
```

also works for newtypes

**Comment by skeate** [comment]
this runs into issues attempting to implement `Guard` since, for some `Iso<A, B>`, `is` should take an `unknown` value and prove it `is B` when all we know is `Guard<A>`. we cannot simply `flow(reverseGet, guardA.is)` the unknown value since we can't really be sure the internal implementation of the `Iso` handles completely irrelevant types in a useful way

**Comment by jacob-alford** [comment]
I’m thinking about renaming this to `WithInvariant` and the property being `imap` because I think technically that’s more accurate.  And I also think that it paints a better picture of what it’s doing, since we basically just need both covariant and contravariant maps simultaneously – which is an invariant functor

**Comment by skeate** [comment]
I think `iso` is the better name. Invariant is a typeclass, so for something to be an invariant functor means it admits an instance of that typeclass; for instance `Maybe` is `Invariant` since you can `invmap intToIntString intStringToInt $ Just 3`.

Indeed [`Iso` itself is invariant](https://gcanti.github.io/monocle-ts/modules/Iso.ts.html#imap) but the important part there is it deals with the higher kinded type `Iso<S, _>`.

In this case we aren't really dealing with higher kinded types; we really just want to be able to convert losslessly between two types `A` and `B`. That's an isomorphism.

**Comment by jacob-alford** [comment]
I'm not sure that I agree.  It seems like WithIso is actually saying: grab the Invariant instance of the applied Schemable, and apply this transformation.  To me Iso makes less sense because we're asserting an isomorphism, we're applying an invariant transformation

**Comment by jacob-alford** [comment]
And we are also invariant mapping HKTs - they're all the Schemable instances, and Schema itself.  The function signature is almost identical.  The one difference is that we have to get the imap by providing a guard and a name

```ts
{
  imap: <A, B>(f: (a: A) => B, g: (b: B) => A) => <E>(i: HKT2<S,O,A>) => HKT2<S,O,B>
}
```

I get that we're not admitting any instances here, but we are invariant mapping using a Schemable's invariant instance (or a super set, like Functor).  So maybe Invariant isn't the right name, but I feel like map needs to be in there somewhere, because we are performing a functor like action and it's clearer to me what's happening.

**Comment by skeate** [comment]
I can see that argument. I'm fine with it being `imap` -- makes `isoDate` less silly anyway

One nitpick though, because I'm me:

> To me Iso makes less sense because we're not asserting an isomorphism

an `Iso` is asserted wherever it is defined; so it's either directly in the argument to `iso` or more likely at some other definition site.

An invariant functor is one where you cannot arbitrarily map its type parameter to some other type. The only way you _can_ is if the two types are isomorphic (as proven by having both functions a->b and b->a, obeying certain laws).

So they are very related concepts; to me the difference is that an `Iso` value is the proof that two types are isomorphic. An invariant constraint is saying that there's some type parameter here that you may only map if you can prove it's isomorphic to the new type. It's more about the container than the data it contains.

Another way of saying it would be that `imap` takes an `iso` as a parameter...

Which I guess means it does make it make sense to name it `imap`. Maybe we can have it take either an existing `Iso` or the functions that would comprise it?

sorry this was rambly

**Comment by skeate** [comment]
> grab the Invariant instance of the applied Schemable

also this framing makes the Guard problem more obvious: it _doesn't have an `Invariant` instance_

**Comment by jacob-alford** [comment]
No that makes sense!  I also thought about having another method which did the same thing and took an iso

**Comment by jacob-alford** [comment]
Or just taking an iso as an argument is good with me too

**Comment by skeate** [comment]
`imap(aToB, bToA)` is _isomorphic to_ `imap({ get: aToB, reverseGet: bToA })`

**Comment by skeate** [comment]
but being able to take an iso makes it handy for the newtype case since then you can just pass in the iso you probably already have

**Comment by jacob-alford** [comment]
Closed with #152
